src/index.js

```diff
@@ -1,9 +1,151 @@
+]);
+
+const SCHEMA = {
+	type: 'object',
+	properties: {
+		iterator: {
+			oneOf: [
+				{
+					type: ['array']
+				},
+				{
+					type: 'object',
+					oneOf: [
+						{
+							properties: {
+								$env: {
+									type: 'string'
+								}
+							},
+							required: ['$env'],
+							additionalProperties: false
+						},
+						{
+							not: {
+								required: ['$env']
+							},
+							additionalProperties: true,
+						}
+					]
+				}
+			]
+		},
+		template: {
+			type: ['array', 'object']
+		}
+	},
+	required: ['iterator', 'template'],
+	additionalProperties: false
+};
+
 class ForEachPlugin {
-	constructor(serverless, options) {
+	constructor(serverless) {
 		this.serverless = serverless;
-		this.options = options;
+
+		this.ajv = new Ajv({ allowUnionTypes: true });
+		this.validator = this.ajv.compile(SCHEMA);
+
+		this.hooks = {
+			'after:package:initialize': this.replace.bind(this),
+		};
+
+		if (this.serverless.configSchemaHandler && this.serverless.configSchemaHandler.defineFunctionEvent) {
+			this.serverless.configSchemaHandler.defineFunctionEvent('aws', '$forEach', SCHEMA);
+		}
+	}
+
+	interpolate(obj, key, value) {
+		const stringified = JSON.stringify(obj);
+		return JSON.parse(stringified
+			.replace('$forEach.key', key)
+			.replace('$forEach.value', value)
+		);
+	}
+
+	findAndReplace(obj, path) {
+		if (Array.isArray(obj)) {
+			obj.forEach((val, idx) => this.findAndReplace(val, `${path}[${idx}]`));
+		} else if (typeof obj === 'object' && obj !== null) {
+			for (const key in obj) {
+				if (key === '$forEach') {
+					this.log(`Found a match in ${path}`);
+
+					this.validate(obj[key], `${path}/${key}`);
+
+					const { iterator: rawIterator, template } = obj[key];
+
+					const iterator = {};
+					if (rawIterator.$env) {
+						Object.entries(process.env).forEach(([name, value]) => {
+							if (name.match(rawIterator.$env)) {
+								iterator[name] = value;
+							}
+						});
+					} else if (Array.isArray(rawIterator)) {
+						rawIterator.forEach((value, idx) => iterator[idx] = value);
+					} else {
+						Object.assign(iterator, rawIterator);
+					}
+
+					const interpolated = Object.entries(iterator).reduce((acc, [key, value]) => {
+						const entry = this.interpolate(template, key, value);
+
+						return Array.isArray(template)
+							? acc.concat(entry)
+							: { ...acc, ...entry };
+					}, Array.isArray(template) ? [] : {});
+
+					const pathIsArray = path.match(/^(.+)\[(\d+)\]$/);
+
+					if (pathIsArray && Array.isArray(interpolated)) {
+						const basePath = pathIsArray[1];
+						const index = Number(pathIsArray[2]);
+
+						const original = get(this.serverless.service, basePath);
+						const newArray = original.slice(0, index).concat(interpolated).concat(original.slice(index + 1));
+
+						set(this.serverless.service, basePath, newArray);
+					} else {
+						const { $forEach, ...result } = obj; // eslint-disable-line no-unused-vars
```

Comment:
This may be moot because serverless doesn't support it. But what if they do something like: 

```yml
events:
  $forEach:
    iterator:
      $env: STREAM_ARN$ # matches all env vars with this suffix
    template:
      - stream:
          arn: $forEach.value
  startingPosition: TRIM_HORIZON
  enabled: true
```

`{ ...result, ...interpolated }` would try and combine an array object with the rest of the properties.

We may want to just add a check and error in theses cases.